lead-time/controller/lead_time_controller.py

```diff
@@ -1,10 +1,134 @@
-from fastapi import APIRouter
+from fastapi import APIRouter, Header
+from typing import Annotated
+
+from pydantic import BaseModel
+from datetime import date
+from library.project.getProjectBySlug import get_project_by_slug
+from library.userStory.get_user_story_history import get_closed_user_stories
 
 router = APIRouter()
 
 
+class TimeRequest(BaseModel):
+    projectslug: str
+    from_date: str = None
+    to_date: str = None
+
+
+class SimpleCache:
+    def __init__(self):
+        self.cache = {}
+
+    def set(self, key, value):
+        self.cache[key] = value
+
+    def get(self, key):
+        return self.cache.get(key)
+
+    def delete(self, key):
+        if key in self.cache:
+            del self.cache[key]
+
+    def clear(self):
+        self.cache.clear()
+
+
+# Create an instance of SimpleCache that can be imported in other files
+cache = SimpleCache()
+
+
 @router.post("/metric/LeadTime")
-def get_lead_time_metric():
+def get_lead_time_metric(
+    lead_time_request: TimeRequest,
+    token: Annotated[str | None, Header()] = None
+):
+    if cache.get(lead_time_request.projectslug) is None:
+        project_info = get_project_by_slug(
+            lead_time_request.projectslug,
+            token
+        )
+        cache.set(lead_time_request.projectslug, project_info)
+        return get_lead_time_details(
+            project_info,
+            token,
+            lead_time_request.from_date,
+            lead_time_request.to_date
+        )
+    else:
+        return get_lead_time_details(
+            cache.get(lead_time_request.projectslug),
+            token,
+            lead_time_request.from_date,
+            lead_time_request.to_date
+        )
+
+
+def get_lead_time_details(project_details,
```

Comment:
All the methods from here till the end, move them to service package which has a file lead_time_service.py

Reply:
Modified